flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamOperatorBase.java

```diff
@@ -0,0 +1,488 @@
+import org.apache.flink.runtime.metrics.groups.TaskManagerJobMetricGroup;
+import org.apache.flink.runtime.metrics.groups.UnregisteredMetricGroups;
+import org.apache.flink.runtime.state.CheckpointStreamFactory;
+import org.apache.flink.runtime.state.KeyedStateBackend;
+import org.apache.flink.runtime.state.OperatorStateBackend;
+import org.apache.flink.runtime.state.StateInitializationContext;
+import org.apache.flink.runtime.state.StateSnapshotContext;
+import org.apache.flink.runtime.state.VoidNamespace;
+import org.apache.flink.runtime.state.VoidNamespaceSerializer;
+import org.apache.flink.streaming.api.graph.StreamConfig;
+import org.apache.flink.streaming.api.watermark.Watermark;
+import org.apache.flink.streaming.runtime.streamrecord.LatencyMarker;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.streaming.runtime.tasks.ProcessingTimeService;
+import org.apache.flink.streaming.runtime.tasks.StreamTask;
+import org.apache.flink.streaming.util.LatencyStats;
+import org.apache.flink.util.Preconditions;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Arrays;
+import java.util.Locale;
+import java.util.Optional;
+
+/**
+ * New base class for all stream operators, replacing previous {@link AbstractStreamOperator}.
+ * Currently intended to work with {@link MultipleInputStreamOperator}.
+ *
+ * <p>One note-able difference in comparison to {@link AbstractStreamOperator} is lack of
+ * {@link AbstractStreamOperator#setup(StreamTask, StreamConfig, Output)} in favor of initialisation
+ * in the constructor, and removed some tight coupling with classes like {@link StreamTask}.
+ *
+ * <p>Methods are guaranteed not to be called concurrently.
+ *
+ * @param <OUT> The output type of the operator
+ */
+@Experimental
+public abstract class StreamOperatorBase<OUT> implements StreamOperator<OUT> {
+	/** The logger used by the operator class and its subclasses. */
+	protected static final Logger LOG = LoggerFactory.getLogger(StreamOperatorBase.class);
+
+	protected final StreamConfig config;
+	protected final Output<StreamRecord<OUT>> output;
+	private final StreamingRuntimeContext runtimeContext;
+	private final ExecutionConfig executionConfig;
+	private final ClassLoader userCodeClassLoader;
+	private final CloseableRegistry cancelables;
+	private final long[] inputWatermarks;
+
+	/** Metric group for the operator. */
+	protected final OperatorMetricGroup metrics;
+	protected final LatencyStats latencyStats;
+	protected final ProcessingTimeService processingTimeService;
+
+	private StreamOperatorStateHandler stateHandler;
+
+	// We keep track of watermarks from both inputs, the combined input is the minimum
+	// Once the minimum advances we emit a new watermark for downstream operators
+	private long combinedWatermark = Long.MIN_VALUE;
+
+	public StreamOperatorBase(StreamOperatorInitializer<OUT> initializer, int numberOfInputs) {
```

Comment:
👍 to no `setup`. I'd really wish we could get rid of `initializeState` as well. Maybe with V3 ;).

(meta: it's hard for me to see the changes to V1. I have seen no obvious problem with the class itself, but I haven't comapred it)